Add OTOT algo

trangmei opened the pull request:
Implements OTOT algo which applies the appropriate algo (either accelerating or decelerating) depending on the student's learning pace. Specs discussed at https://groups.google.com/d/topic/coursemology/Hu24Fjf-rlE/discussion
Fixes https://github.com/Coursemology/coursemology2/issues/3686

Add checks to prevent past personal times from being moved. Fixes https://github.com/Coursemology/coursemology2/issues/3688

fonglh commented on app/controllers/concerns/course/lesson_plan/personalization_concern.rb:
```diff
@@ -28,6 +28,26 @@ def algorithm_fixed(course_user)
+      algorithm_fomo(course_user)
+    else
+      algorithm_stragglers(course_user)
+    end
```
`learning_rate_ema < 1 ? algorithm_fomo(course_user) : algorithm_stragglers(course_user)`

fonglh commented on app/controllers/concerns/course/lesson_plan/personalization_concern.rb:
```diff
@@ -71,11 +91,13 @@ def algorithm_fomo(course_user)
             course_tz,
             FOMO_DATE_ROUNDING_THRESHOLD
-          )
+          ) if personal_time.start_at > Time.zone.now
```
Should this be based on the user's timezone or the course timezone instead?

For an example of course timezone use, see https://github.com/Coursemology/coursemology2/blob/master/app/jobs/consolidated_item_email_job.rb#L10

fonglh replied:
Same for the other `Time.zone.now` comparisons below.

trangmei replied:
This should be comparison between two datetimes with zone, so any difference in timezone should be handled by rails.

trangmei commented:
Tested on local with prod data

1. Switched a leaderboard student from FOMO to OTOT, nothing changes. This is correct because the fast student's timeline is essentially still subjected to FOMO calculation.

2. Switched a student with learning rate > 1 (randomly picked a student who submitted an overdue mission) from FOMO to OTOT:

Before:
![image](https://user-images.githubusercontent.com/35135264/67194062-f9e64e00-f428-11e9-9e4a-c873decc0788.png)

After:
![image](https://user-images.githubusercontent.com/35135264/67194091-08cd0080-f429-11e9-8cf1-34770beeb67b.png)

This is correct because slow student's timeline is now subjected to Stragglers calculation, and we see deadline extension.

3. Switched a leaderboard student from FOMO to Straggler's algo. Assessments that were opened before remains open. Assessments that were not yet opened on FOMO get reset to reference opening date.

This is correct because personal times in the past should not change at all.
